File: Java-and-C++/JavaUDx/UDAnalytics/com/vertica/JavaLibs/Lag.java
/* Copyright (c) 2005 - 2016 Hewlett Packard Enterprise Development LP  -*- JAVA -*-*/
/* 
 * Description: Example User Defined Analytic Function Lag.
 *
 * Create Date: Jan 7 2015
 */
package com.vertica.JavaLibs;

import java.util.ArrayList;

import com.vertica.sdk.AnalyticFunction;
import com.vertica.sdk.AnalyticPartitionReader;
import com.vertica.sdk.AnalyticPartitionWriter;
import com.vertica.sdk.DestroyInvocation;
import com.vertica.sdk.ServerInterface;
import com.vertica.sdk.SizedColumnTypes;
import com.vertica.sdk.UdfException;

/*
 * Lag takes in an input expression and an offset value which determines how far to
 * lag behind the current row, and returns the expression evaluated on that row.
 * This function mimics the built-in analytic function LAG, but
 * it is implemented only for integer input expressions.
 */

public class Lag extends AnalyticFunction {

	@Override
	public void processPartition(ServerInterface srvInterface,
			AnalyticPartitionReader inputReader, AnalyticPartitionWriter outputWriter)
			throws UdfException, DestroyInvocation {
		
		SizedColumnTypes inTypes = inputReader.getTypeMetaData();
		ArrayList<Integer> argCols = new ArrayList<Integer>();
		
		inTypes.getArgumentColumns(argCols);
		
		long  lagOffset = inputReader.getLong(argCols.get(1)); // the offset
		long currentOffset = 0;
		long value;
		
		do{
			if(currentOffset == lagOffset){
				value = inputReader.getLong(argCols.get(0)); // get lagged value.
				outputWriter.setLong(0, value);
				inputReader.next();
			}else {
				++currentOffset;
				outputWriter.setLongNull(0);
			}
		}while(outputWriter.next());
	}

}

File: Java-and-C++/AnalyticFunctions/Lag.cpp
/* Copyright (c) 2005 - 2016 Hewlett Packard Enterprise Development LP  -*- C++ -*-*/
/* 
 * Description: Example User Defined Analytic Function Lag.
 *
 * Create Date: Nov 22, 2011
 */
#include "Vertica.h"
#include <sstream>

using namespace Vertica;

/*
 * Lag takes in an input expression and an offset value which determines how far to
 * lag behind the current row, and returns the expression evaluated on that row.
 * This function mimics the built-in analytic function LAG, but
 * it is implemented only for integer input expressions.
 */
class Lag : public AnalyticFunction
{
    virtual void processPartition(ServerInterface &srvInterface, 
                                  AnalyticPartitionReader &inputReader, 
                                  AnalyticPartitionWriter &outputWriter)
    {
        try {
            const SizedColumnTypes &inTypes = inputReader.getTypeMetaData();
            std::vector<size_t> argCols; // Argument column indexes.
            inTypes.getArgumentColumns(argCols);

            vint lagOffset = inputReader.getIntRef(argCols.at(1)); // the offset.
            vint currentOffset = 0;
            vint value;

            do {
                if (currentOffset == lagOffset) {
                    value = inputReader.getIntRef(argCols.at(0)); // get lagged value.
                    outputWriter.setInt(0, value);
                    inputReader.next();
                }
                else {
                    ++currentOffset;
                    outputWriter.setInt(0, vint_null);
                }
            } while (outputWriter.next());
        } catch(std::exception& e) {
            // Standard exception. Quit.
            vt_report_error(0, "Exception while processing partition: [%s]", e.what());
        }
    }
};

class LagFactory : public AnalyticFunctionFactory
{
    virtual void getPrototype(ServerInterface &srvInterface, ColumnTypes &argTypes, ColumnTypes &returnType)
    {
        argTypes.addInt();
        argTypes.addInt();
        returnType.addInt();
    }

    virtual void getReturnType(ServerInterface &srvInterface, 
                               const SizedColumnTypes &inputTypes, 
                               SizedColumnTypes &outputTypes)
    {
        outputTypes.addInt();
    }

    virtual AnalyticFunction *createAnalyticFunction(ServerInterface &srvInterface)
    { return vt_createFuncObject<Lag>(srvInterface.allocator); }

};

RegisterFactory(LagFactory);
